### PR TITLE
Nc 2107 permissioning config toml file

### PR DESCRIPTION
**Pull request by macfarla**
## PR description
Parse TOML config file for permissions config options
Also added CLI options
--permissions-nodes-enabled
--permissions-accounts-enabled

## Fixed Issue(s)
NC-2107

### CODE REVIEW

File: pantheon/src/main/java/tech/pegasys/pantheon/util/PermissioningConfigurationValidator.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+              .stream()
+              .map(DefaultPeer::fromURI)
+              .collect(Collectors.toList());
+      for (final Peer bootnode : bootnodes) {
+        if (!whitelist.contains(bootnode)) {
+          bootnodesNotInWhitelist.add(bootnode);
+        }
+      }
+      if (!bootnodesNotInWhitelist.isEmpty()) {
+        throw new Exception("Bootnode(s) not in nodes-whitelist " + bootnodesNotInWhitelist);
```
[the remark this answers is not in the document]
Comment (mark-terry) [reply pantheon/src/main/java/tech/pegasys/pantheon/util/PermissioningConfigurationValidator.java]: I've got a change in my NC-1968 branch that moves the Peer -> enode URI functionality into the Peer implementations, which is probably what we need to use here.

[the remark this answers is not in the document]
Comment (macfarla) [reply pantheon/src/main/java/tech/pegasys/pantheon/util/PermissioningConfigurationValidator.java]: yes you are absolutely right. This made it compile but I didn't get back to fixing it properly.

File: pantheon/src/main/java/tech/pegasys/pantheon/TomlConfigFileParser.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+      final TomlParseResult result, final String toml) {
+    if (result == null || result.isEmpty()) {
+      LOG.error("Unable to read TOML configuration file %s", toml);
+    }
+    return result;
+  }
+
+  public static TomlParseResult loadConfiguration(final String toml) {
+    TomlParseResult result;
+    result = Toml.parse(toml);
```
[the remark this answers is not in the document]
Comment (macfarla) [reply pantheon/src/main/java/tech/pegasys/pantheon/TomlConfigFileParser.java]: 👍

File: pantheon/src/main/java/tech/pegasys/pantheon/PermissioningConfigurationBuilder.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    }
+  }
+
+  private static File permissioningConfigFile(final String filename) {
+
+    final File permissioningConfigFile = new File(filename);
+    if (permissioningConfigFile.exists()) {
+      return permissioningConfigFile;
+    } else {
+      LOG.error("Unable to load permissioning config file from location {}.", filename);
```
[the remark this answers is not in the document]
Comment (macfarla) [reply pantheon/src/main/java/tech/pegasys/pantheon/PermissioningConfigurationBuilder.java]: 👍

File: pantheon/src/main/java/tech/pegasys/pantheon/util/PermissioningConfigurationValidator.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        DiscoveryConfiguration.getBootstrapPeersFromGenericCollection(
+            ethNetworkConfig.getBootNodes());
+    if (permissioningConfiguration.isNodeWhitelistSet() && bootnodes != null) {
+      final List<Peer> whitelist =
+          permissioningConfiguration
+              .getNodeWhitelist()
+              .stream()
+              .map(DefaultPeer::fromURI)
+              .collect(Collectors.toList());
+      for (final Peer bootnode : bootnodes) {
```
[the remark this answers is not in the document]
Comment (macfarla) [reply pantheon/src/main/java/tech/pegasys/pantheon/util/PermissioningConfigurationValidator.java]: 👍

File: pantheon/src/main/java/tech/pegasys/pantheon/util/PermissioningConfigurationValidator.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (macfarla) [reply pantheon/src/main/java/tech/pegasys/pantheon/util/PermissioningConfigurationValidator.java]: temporarily copied buildEnodeURI from PermGetNodesWhitelist (will be cleaned up by your change in NC-1968)

**Pull request merged**